File: platforms/android/keyboard-renderer/src/main/java/app/funput/funput/keyboard/model/KeyboardLayout.kt
package app.funput.funput.keyboard.model

data class KeySpec(
    val id: String,
    val label: String,
    val role: KeyRole,
    val widthWeight: Float = 1f,
    val shiftedLabel: String? = null,
    val secondaryLabel: String? = null,
    val accessibilityLabel: String = label,
    val horizontalSwipeAction: KeySwipeAction? = null,
) {
    init {
        require(id.isNotBlank()) { "Key id must not be blank" }
        require(widthWeight > 0f) { "Key width weight must be positive" }
        require(shiftedLabel == null || shiftedLabel.isNotEmpty()) { "Shifted label must not be empty" }
        require(accessibilityLabel.isNotBlank()) { "Accessibility label must not be blank" }
    }
}

enum class KeySwipeAction {
    TOGGLE_LANGUAGE,
}

data class KeyboardRow(
    val keys: List<KeySpec>,
    val horizontalInsetUnits: Float = 0f,
) {
    init {
        require(keys.isNotEmpty()) { "Keyboard row must contain at least one key" }
        require(horizontalInsetUnits >= 0f) { "Row inset must not be negative" }
        require(keys.map(KeySpec::id).distinct().size == keys.size) { "Key ids must be unique within a row" }
    }
}

data class SuggestionBarSpec(
    val settingsKey: KeySpec,
    val emojiKey: KeySpec,
    val suggestionsEnabled: Boolean = true,
) {
    init {
        require(settingsKey.role == KeyRole.SETTINGS) { "Toolbar settings action must be a settings key" }
        require(emojiKey.role == KeyRole.EMOJI) { "Toolbar emoji action must be an emoji key" }
    }
}

data class KeyboardLayout(
    val id: String,
    val inputMethod: KeyboardInputMethod,
    val suggestionBar: SuggestionBarSpec?,
    val rows: List<KeyboardRow>,
) {
    init {
        require(id.isNotBlank()) { "Layout id must not be blank" }
        require(rows.isNotEmpty()) { "Keyboard layout must contain at least one row" }

        val keyIds = buildList {
            suggestionBar?.settingsKey?.let { add(it.id) }
            suggestionBar?.emojiKey?.let { add(it.id) }
            rows.forEach { row -> addAll(row.keys.map(KeySpec::id)) }
        }
        require(keyIds.distinct().size == keyIds.size) { "Key ids must be unique within a layout" }
    }
}

File: platforms/ios/Packages/FunputKit/Sources/KeyboardLayout/Models/KeyboardLayout.swift
import Foundation

public struct KeyboardRow: Hashable, Sendable {
    public let keys: [KeySpec]
    public let horizontalInsetUnits: CGFloat

    public init(keys: [KeySpec], horizontalInsetUnits: CGFloat = 0) {
        precondition(!keys.isEmpty, "Keyboard row must contain at least one key")
        precondition(horizontalInsetUnits >= 0, "Row inset must not be negative")

        self.keys = keys
        self.horizontalInsetUnits = horizontalInsetUnits
    }
}

public struct KeyboardLayout: Hashable, Sendable {
    public let id: String
    public let inputMethod: KeyboardInputMethod
    public let toolbar: KeyboardToolbarSpec?
    public let rows: [KeyboardRow]

    public init(
        id: String,
        inputMethod: KeyboardInputMethod = .telex,
        toolbar: KeyboardToolbarSpec?,
        rows: [KeyboardRow]
    ) {
        precondition(!id.isEmpty, "Keyboard layout id must not be empty")
        precondition(!rows.isEmpty, "Keyboard layout must contain at least one row")

        let ids = (toolbar?.keys.map(\.id) ?? []) + rows.flatMap(\.keys).map(\.id)
        precondition(Set(ids).count == ids.count, "Key ids must be unique within a layout")

        self.id = id
        self.inputMethod = inputMethod
        self.toolbar = toolbar
        self.rows = rows
    }

    /// Whether this layout gives the user an explicit route into Funput's emoji panels.
    public var allowsEmojiPanel: Bool {
        toolbar?.keys.contains { $0.role == .emoji } == true
            || rows.contains { row in row.keys.contains { $0.role == .emoji } }
    }
}
